code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/load/Unifier.kt

```diff
@@ -0,0 +1,91 @@
+import org.jetbrains.research.ml.ast.transformations.ifRedundantLinesRemoval.IfRedundantLinesRemovalTransformation
+import org.jetbrains.research.ml.ast.transformations.multipleOperatorComparison.MultipleOperatorComparisonTransformation
+import org.jetbrains.research.ml.ast.transformations.multipleTargetAssignment.MultipleTargetAssignmentTransformation
+import org.jetbrains.research.ml.ast.transformations.outerNotElimination.OuterNotEliminationTransformation
+import org.jetbrains.research.ml.ast.util.sdk.setSdkToProject
+
+private const val MAX_ITERATIONS: Int = 5
+
+/**
+ * Class for unifying code solutions using ast transformations.
+ */
+class Unifier {
+    private val project: Project = ProjectUtil.openOrImport(getTmpProjectDir(), null, true)
+        ?: error("Internal error: the temp project was not created")
+    private val psiManager: PsiManager = PsiManager.getInstance(project)
+    private val allTransformations: List<Transformation> = listOf(
+        AnonymizationTransformation,
+        AugmentedAssignmentTransformation,
+        CommentsRemovalTransformation,
+        ComparisonUnificationTransformation,
+        ConstantFoldingTransformation,
+        DeadCodeRemovalTransformation,
+        ExpressionUnificationTransformation,
+        IfRedundantLinesRemovalTransformation,
+        MultipleOperatorComparisonTransformation,
+        MultipleTargetAssignmentTransformation,
+        OuterNotEliminationTransformation,
+    )
+
+    init {
+        setSdkToProject(project, getTmpProjectDir())
+    }
+
+    private fun String.createPsiFile(id: Int): PsiFile = ApplicationManager.getApplication().runWriteAction<PsiFile> {
+        val basePath = getTmpProjectDir(toCreateFolder = false)
+        val fileName = "dummy$id.${PythonFileType.INSTANCE.defaultExtension}"
+        val file = addPyFileToProject(basePath, fileName, fileContext = this)
+        val virtualFile = LocalFileSystem.getInstance().refreshAndFindFileByIoFile(file)
+        psiManager.findFile(virtualFile!!)
+    }
+
+    private fun PsiFile.applyTransformations(transformations: List<Transformation>) {
+        ApplicationManager.getApplication().invokeAndWait {
+            ApplicationManager.getApplication().runWriteAction {
+                transformations.forEach {
+                    it.forwardApply(this)
+                }
+            }
+        }
+    }
+
+    private fun PsiFile.hasChangedAfterTransformations(transformations: List<Transformation>): Boolean {
+        val prevPsi = this.copy()
+        this.applyTransformations(transformations)
+        return prevPsi.text != this.text
+    }
+
+    fun Submission.unify(): Submission {
+        val psi = this.code.createPsiFile(this.id)
+
+        var iterationsCounter = 0
+        do {
+            iterationsCounter += 1
+            if (!psi.hasChangedAfterTransformations(allTransformations)) {
+                break
+            }
+        } while (iterationsCounter <= MAX_ITERATIONS)
```

Comment:
I think it would be clearer if we union the conditions:
```
        var wasChanged: Boolean
        var iterationsCounter = 0
        do {
            wasChanged = psi.hasChangedAfterTransformations(allTransformations)
            iterationsCounter += 1
        } while (wasChanged || iterationsCounter <= MAX_ITERATIONS)
```